telegram/ext/_fsm/machine.py

```diff
@@ -0,0 +1,200 @@
+        the current version, the state should not be updated.
+
+        Important:
+            This should be used exclusively by methods of this class and subclasses.
+            It should *not* be called directly by users of this class!
+        """
+        _LOGGER.debug("Setting %s state to %s", key, state)
+        if state is State.ANY:
+            raise ValueError("State.ANY is not supported in set_state")
+
+        if version and version != self._states.get(key, (None, None))[1]:
+            raise ValueError("Optimistic locking failed. Not updating state.")
+
+        if jq := self._get_job_queue(raise_exception=False):
+            # This is a rather tight coupling between FSM and JobQueue
+            # Not sure if we like that. Makes it even harder to replace JobQueue
+            # (or the JQ implementation) with something else.
+            # The upside is that we don't need to maintain any additional internal state
+            # for the jobs and persistence is handled by the JobQueue.
+            cancel_jobs = jq.jobs(pattern=str(hash(key)))
+            for job in cancel_jobs:
+                _LOGGER.debug("Cancelling timeout job %s", job)
+                job.schedule_removal()
+
+        # important to use time.perf_counter_ns() here, as time_ns() is not monotonic
+        self.__states[key] = (state, time.perf_counter_ns())
+        # Doing this *after* do_set_state so that any exceptions are raised before the history
+        # is updated
+        self.store_state_history(key, state)
+        return StateInfo(key, state, self._states[key][1])
+
+    async def set_state(self, key: _KT, state: State, version: Optional[int] = None) -> None:
+        """Store the state for the specified key."""
+        async with self.get_lock(key):
+            self._do_set_state(key, state, version)
+
+    def set_state_nowait(self, key: _KT, state: State, version: Optional[int] = None) -> None:
+        """Store the state for the specified key without waiting for a lock."""
+        if self.get_lock(key).locked():
+            raise asyncio.InvalidStateError("Lock is locked")
+        self._do_set_state(key, state, version)
+
+    @contextlib.asynccontextmanager
+    async def as_state(self, key: _KT, state: State) -> AsyncIterator[None]:
+        """Context manager to set the state for the specified key and reset it afterwards."""
+        async with self.get_lock(key):
+            current_state, current_version = self.states[key]
+            new_version = self._do_set_state(key, state, current_version).version
+            try:
+                yield
+            finally:
+                self._do_set_state(key, current_state, new_version)
+
+    @staticmethod
+    def _build_job_name(keys: Sequence[_KT]) -> str:
+        return f"FSM_Job_{'_'.join(str(hash(k)) for k in keys)}"
+
+    def set_job_queue(self, job_queue: "JobQueue") -> None:
+        self.__job_queue = weakref.ref(job_queue)
```

Comment:
Why weakref wont it/the user experience break if the supplied job queue gets destroyed

Reply:
There is `job_queue.application`. that's also weakref b/c we have a cyclic reference: `application is application.job_queue.application`. Now we also have `application is application.fsm.job_queue.application`. I'm not 100% sure if this is strictly necessary, but it seemed safer to me 🤔 I can at very least add a comment.